pkg/controller/cluster_pools_pull_secret_provider/cluster_pools_pull_secret_provider.go

```diff
@@ -0,0 +1,210 @@
+		secretHandler(sourcePullSecretNamespace, sourcePullSecretName, hiveClient, requestsFactoryForSecretEvent),
+	); err != nil {
+		return fmt.Errorf("failed to create watch for secrets: %w", err)
+	}
+
+	r.log.Info("Successfully added reconciler to manager")
+	return nil
+}
+
+func clusterPoolHandler(sourcePullSecretNamespace string) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(o ctrlruntimeclient.Object) []reconcile.Request {
+		clusterPool, ok := o.(*hivev1.ClusterPool)
+		if !ok {
+			logrus.WithField("type", fmt.Sprintf("%T", o)).Error("Got object that was not a ClusterPool")
+			return nil
+		}
+		if clusterPool.Namespace == sourcePullSecretNamespace {
+			return nil
+		}
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{
+			Namespace: clusterPool.Namespace,
+			Name:      clusterPool.Name,
+		}}}
+	})
+}
+
+func requestsFactoryForSecretEvent(namespace string, client ctrlruntimeclient.Client) []reconcile.Request {
+	if namespace != "" {
+		clusterPools := &hivev1.ClusterPoolList{}
+		if err := client.List(context.TODO(), clusterPools, ctrlruntimeclient.InNamespace(namespace)); err != nil {
+			logrus.WithField("namespace", namespace).WithError(err).Error("Failed to list the cluster pools in namespace")
+		}
+		if len(clusterPools.Items) == 0 {
+			logrus.WithField("namespace", namespace).Info("No cluster pools are in namespace")
+			return nil
+		}
+		logrus.WithField("namespace", namespace).WithField("name", clusterPools.Items[0].Name).Info("Found cluster pool")
+		return []reconcile.Request{{NamespacedName: types.NamespacedName{
+			Namespace: namespace,
+			Name:      clusterPools.Items[0].Name,
+		}}}
+	}
+	var requests []reconcile.Request
+	clusterPools := &hivev1.ClusterPoolList{}
+	if err := client.List(context.TODO(), clusterPools); err != nil {
+		logrus.WithError(err).Error("Failed to list the cluster pools")
+	}
+	namespacesWithClusterPool := sets.NewString()
+	for _, p := range clusterPools.Items {
+		if namespacesWithClusterPool.Has(p.Namespace) {
+			continue
+		}
+		namespacesWithClusterPool.Insert(p.Namespace)
+		logrus.WithField("namespace", p.Namespace).WithField("name", p.Name).Info("Found cluster pool")
+		requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+			Namespace: p.Namespace,
+			Name:      p.Name,
+		}})
+	}
+	return requests
+}
+
+func secretHandler(
+	sourcePullSecretNamespace string,
+	sourcePullSecretName string,
+	client ctrlruntimeclient.Client,
+	requestsFactory func(namespace string, client ctrlruntimeclient.Client) []reconcile.Request,
+) handler.EventHandler {
+	return handler.EnqueueRequestsFromMapFunc(func(o ctrlruntimeclient.Object) []reconcile.Request {
+		secret, ok := o.(*corev1.Secret)
+		if !ok {
+			logrus.WithField("type", fmt.Sprintf("%T", o)).Error("Got object that was not a Secret")
+			return nil
+		}
+		if secret.Name != sourcePullSecretName {
```

Comment:
See the other comment re namespace, this can just be
```suggestion
		if secret.Name != sourcePullSecretName || secret.Namespace == sourcePullSecret Namespace {
```

allowing you to simplify the code below and only call `requestsFactory` once

Reply:
Still need to call it twice because I still look at pool in reconcile.